### ISSUE TITLE
Replace generic syntax errors with more informative messages

### ISSUE THREAD

**Issue by bradcray** [chapel-lang/chapel#17467]
### Summary of Problem

As a new (or even seasoned) Chapel programmer, I'll inevitably make some simple mistakes in my code that will correctly get reported as syntax errors by the compiler.  As a trivial example, if I were to write:

```chapel
proc foo() {
  writeln("In foo");
// note the missing curly bracket here
```

I get:

```
tesit2.chpl:4: syntax error
```

which is accurate but not very helpful.  In a short program like this, it may not be too difficult to determine what's wrong, but in a much larger program...

This issue is a placeholder for generally improving such situations.

### Configuration Information

- Output of `chpl --version`: `1.24.0`

**Comment by vrockz747** [comment]
Hey @bradcray, I would like to work on this issue.
Please guide me as to how to get started.

**Comment by bradcray** [comment]
Hi @vrockz747 — Thanks for your enthusiasm for helping out here.  I actually don't feel expert enough in what needs to occur to know how to advise you very specifically.  What I know is:  We use flex + bison (standard parser generator tools) for our parser (the files are in compiler/parser/chapel.lex and compiler/parser/chapel.ypp), and that it's the failure to match existing rules in bison that leads to the generic syntax errors.  But I'm not expert enough in bison to have an idea for how we might go about improving them.  I suspect that this is a common problem that many bison-based parsers run into, so doing a Google search may give you ideas (or may make you decide that abandoning bison is the only sane way to go).  PR #11405 may be good for inspiration as well.  IIRC, it turned cases like `var int: bool;` from syntax errors into more helpful errors (where `int` is a reserved word in Chapel, so can't be used as a new variable's name).

**Comment by vrockz747** [comment]
Thank you for your quick response @bradcray, I am researching a bit on bishop and flex and will try to cover this issue(I didn't think it would be this complex at first :D ).

**Comment by bradcray** [comment]
I don't think it's trivial, but I'm not sure how complex it is either (though I do think it's somewhat boundless given that there are arbitrarily many syntax errors one can run into).  No shame if you decide this isn't an issue where you'd want to spend your time.

**Comment by vrockz747** [comment]
@bradcray, While researching a bit I found that adding ```%define parse.error verbose``` [(source)](https://stackoverflow.com/questions/59467072/find-where-syntax-error-in-flex-yacc-happened)  to the bison file(compiler/parser/chapel.ypp) will change the error message to something more helpful than just "syntax error", and I found these lines commented in the file:
```
// These would print out expected token names, but that might just
// add unnecessary complexity for someone new to the language ...
%define parse.lac full //this line was commented originally
%define parse.error verbose //this line was commented originally
```
but even after uncommenting these lines and recompiling them nothing seems to happen.
I suppose since these lines were originally written in the code(and purposely commented for the convenience of new users)
I think there might be something very trivial I am missing here.

**Comment by bradcray** [comment]
> but even after uncommenting these lines and recompiling them nothing seems to happen.

Something that may not be obvious:  In the Chapel build system, an explicit `make parser` step is required to re-build the parser, followed by a normal `make` step to re-build the compiler as a whole.  So that could be why changing something like this doesn't seem to have any impact?

**Comment by vrockz747** [comment]
Oh, I see, I read the ```read me``` file where it was written for the additional ```make parser``` step but was I was skipping the obvious one ```make``` to re-build the compiler :sweat_smile: , after doing that the change is reflected. Now the error is
```2: syntax error, unexpected $end```
I guess now I will try to make this more understandable for users.

**Comment by vrockz747** [comment]
The steps mentioned in the readme:
```
This means that we require that developers who work on the parser:
1) have the following versions installed: Bison 2.5 or later; Flex 2.6 or later
2) explicitly build the generated parser when they alter the parser
sources.  One way to do this is to run "make parser" in this
directory.
3) when submitting a pull request to modify the parser sources, also include
the generated parser.  We find it helpful to add the sources and the
generated files as separate commits within the same pull request.
This enables the reviewer to more easily ignore the extensive diffs
that may appear in the generated parser.
```
I think step 2 can be modified to this :
```
2) explicitly build the generated parser when they alter the parser
sources.  One way to do this is to run "make parser" in this
directory and then "make" step to re-build the compiler as a whole.
```
Is this clarifying the steps correctly?
Should I commit these changes? 
Sorry if I am asking stupid questions I am still learning about these :)

**Comment by bradcray** [comment]
Hi @vrockz747 : I think that edit makes sense and would be beneficial to file as a PR, yeah.  It's also possible to run `make parser` from the top-level $CHPL_HOME directory (which is what I normally do), so it could make sense to remove the bit about "in this directory" and rewrite the bullet to talk in terms of running both commands from $CHPL_HOME to simplify and avoid confusion about where each should be run.

> Sorry if I am asking stupid questions I am still learning about these :)

Please don't worry about that at all.

**Comment by vrockz747** [comment]
> it could make sense to remove the bit about "in this directory"

Yes surely, even I was especially changing my directory to that location before running the command, didn't knew it would do the same when done from $CHPL_HOME directory.
Then I will re-write step2 as : 
```
2) Explicitly build the generated parser when they alter the parser
sources. One way to do this is to run "make parser" and then run "make" to re-build the compiler as a whole.
```

**Comment by tcarneirop** [comment]
Hello Everyone. I was about to add this issue and Elliot pointed to this discussion.

Here is my example:

```
when "a"{
    select var{
            when "b"{
             .....
          }
    }
}
```
In this case, I forgot the ''select" before "when a". I receive ```syntax error near line ...```, which sometimes does not help.

Thank you!
